File: graphs/200_number_of_island.py
def num_of_islands_no_space(grid: list[list[str]]) -> int:
    m = len(grid)
    n = len(grid[0])

    result = 0

    def is_valid(r: int, c: int):
        return 0 <= r < m and 0 <= c < n and grid[r][c] == "1"

    for i in range(m):
        for j in range(n):
            if grid[i][j] == "1":
                result += 1
                grid[i][j] = "0"
                stack = [(i, j)]
                while stack:
                    r, c = stack.pop()

                    for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                        row = r + dx
                        col = c + dy
                        if is_valid(row, col):
                            grid[row][col] = "0"
                            stack.append((row, col))

    return result

def num_of_islands_with_extra_space(grid: list[list[str]]) -> int:
    m = len(grid)
    n = len(grid[0])

    result = 0
    visited = set()

    def is_valid(r, c):
        return 0 <= r < m and 0 <= c < n and grid[r][c] == "1" and (r, c) not in visited

    for i in range(m):
        for j in range(n):
            if grid[i][j] == "1" and (i, j) not in visited:
                result += 1
                visited.add((i, j))
                stack = [(i, j)]
                while stack:
                    r, c = stack.pop()

                    for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                        row = r + dx
                        col = c + dy
                        if is_valid(row, col):
                            visited.add((row, col))
                            stack.append((row, col))

    return result

File: graphs/200_number_of_island.go
// Так хранятся координаты суши
type Tuple struct {
	y, x int
}

func numIslands(grid [][]byte) int {
	result := 0
	m := len(grid)
	n := len(grid[0])

	isValid := func(y, x int) bool {
		return (y >= 0 &&
			y < m &&
			x >= 0 &&
			x < n &&
			grid[y][x] == '1')
	}

	directions := []Tuple{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == '0' {
				continue
			}
			result++
			stack := NewStack[Tuple]()
			stack.Append(Tuple{i, j})
			for len(stack) > 0 {
				tuple := stack.Pop()

				for _, dir := range directions {
					dx := dir.x + tuple.x
					dy := dir.y + tuple.y
					if isValid(dy, dx) {
						grid[dy][dx] = '0'
						stack.Append(Tuple{dy, dx})
					}
				}
			}
		}
	}

	return result
}

// TYPES
type Stack[T any] []T

// удаляет и возвращает последний элемент стека
func (s *Stack[T]) Pop() T {
	pop := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return pop
}

// возвращает последний элемент стека
func (s Stack[T]) Top() T {
	return s[len(s)-1]
}

// добавляет элемент в стэк
func (s *Stack[T]) Append(val T) {
	*s = append(*s, val)
}

func NewStack[T any]() Stack[T] {
	return Stack[T]{}
}

